src/usb.c

```diff
@@ -0,0 +1,1619 @@
+                        GDBusMethodInvocation *invocation,
+                        const char            *arg_parent_window,
+                        GVariant              *arg_devices,
+                        GVariant              *arg_options)
+{
+  g_autoptr(XdpDbusImplRequest) impl_request = NULL;
+  g_autoptr(UsbSenderInfo) sender_info = NULL;
+  g_autoptr(GVariant) filtered_devices = NULL;
+  g_autoptr(GVariant) options = NULL;
+  g_autoptr(GError) error = NULL;
+  GVariantBuilder options_builder;
+  Permission permission;
+  Request *request;
+  Usb *self;
+
+  self = (Usb *) object;
+  request = request_from_invocation (invocation);
+
+  g_debug ("[usb] Handling AccessDevices");
+
+  REQUEST_AUTOLOCK (request);
+
+  permission = get_permission_sync (xdp_app_info_get_id (request->app_info),
+                                    PERMISSION_TABLE,
+                                    PERMISSION_ID);
+  if (permission == PERMISSION_NO)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                             "Not allowed to create USB sessions");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  impl_request = xdp_dbus_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (usb_impl)),
+                                                       G_DBUS_PROXY_FLAGS_NONE,
+                                                       g_dbus_proxy_get_name (G_DBUS_PROXY (usb_impl)),
+                                                       request->id,
+                                                       NULL,
+                                                       &error);
+  if (!impl_request)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  if (!xdp_filter_options (arg_options,
+                           &options_builder,
+                           usb_access_devices_options,
+                           G_N_ELEMENTS (usb_access_devices_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+  options = g_variant_builder_end (&options_builder);
+
+  sender_info = usb_sender_info_from_request (self, request);
+  g_assert (sender_info != NULL);
+
+  if (sender_info->sender_state != USB_SENDER_STATE_DEFAULT)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_FAILED,
+                                             "Cannot call AcquireDevices() with an unfinished "
```

Comment:
Why not? The interface description doesn't mention this restriction at all.

Reply:
The changes I proposed on AcquireDevices would resolve this thread.

Reply:
this is done in the current code.